### PR TITLE
Add Interacting Multiple Model Filter

### PR DESCRIPTION
**Pull request by vkorotkine**
Added interacting multiple model filter. As examples, right now have process model noise estimation. 
Issues with this PR right now:
- Duplication of the ExtendedKalmanFilter into ExtendedKalmanFilterIMM. This is because the IMM needs innovation and innovation covariance which the vanulla EKF class does not return. So theres a good chunk of code duplication (the entire correct method). If anyone has an idea of how to do this better, would appreciate. 
- For now, only implemented for right perturbation.

### CODE REVIEW

**Comment by CharlesCossette** [comment]
Can we please do type hinting on all function arguments?

**Comment by CharlesCossette** [comment]
In the future, lets stop committing documentation build

File: pynav/imm.py [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+        g = X_par.group
+        if X_par.direction == "right":
+            for X in X_list: 
+                mu = g.Log(g.inverse(X_par.value) @ X.state.value)
+                Jr_inv = g.right_jacobian(mu)
+                Sigma = Jr_inv @ X.covariance @ Jr_inv.T 
+                mu = VectorState(mu, stamp = X_par.stamp)
+                x_reparametrized.append(StateWithCovariance(mu, Sigma))
+
+        return x_reparametrized
```
Comment (CharlesCossette) [inline pynav/imm.py:153]:
```suggestion
       x_reparameterized = [] 
       for x in x_list:
           mu = x_par.minus(x)
           J = x_par.jacobian(mu)
           Sigma = J @ x.covariance @ J.T
           x_reparametrized.append(StateWithCovariance(mu, Sigma))
           
       x_reparameterized
```

should work for both vector space, liegroup (left/right), and any other state definition

File: pynav/imm.py [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+        # Add this fudge factor to get through those cases.
+        if np.allclose(mu_k, np.zeros(mu_k.shape)):
+            mu_k = 1e-10 * np.ones(mu_k.shape)
+
+        mu_k = mu_k/np.sum(mu_k)  
+
+        return x_hat, mu_k
+
+
+def run_time_varying_Q_filter(
```
Comment (CharlesCossette) [inline pynav/imm.py:323]: I vote to remove this and handle time-varying Q the more natural way with our framework

File: examples/ex_interacting_multiple_model_se3.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+import numpy as np
+from typing import List
+import time
+from matplotlib import pyplot as plt
+from pynav.imm import InteractingModelFilter, run_interacting_multiple_model_filter
+from pynav.imm import run_time_varying_Q_filter
+from pynav.imm import ImmResultList
+
+
+def make_range_models_iterated_ekf(R):
```
Comment (CharlesCossette) [inline examples/ex_interacting_multiple_model_se3.py:25]: any way to make these examples simpler, making them more like the other ones?

Examples should be a "quickstart"

File: pynav/imm.py (hunk 2 above)
Comment (vkorotkine) [reply pynav/imm.py:323]: I do handle the time-varying Q in our framework. Its just that to run a filter which assumes a time-varying Q it needs to be reinitialized at every step, which is not the case in current run_filter method

Comment (vkorotkine) [reply pynav/imm.py:323]: Ah right just saw the time-varying example update. Yeah this will work.

**Review by CharlesCossette — APPROVED** [review]

**Pull request merged**